---
title: "Working with DNA sequences and features in R with Bioconductor - version 2"
output:
    html_document:
        theme: cerulean
---

## Workshop material

* [Slides](slides.html)
* [Workshop](workshop.html)

## Setup

Download and install:

* [R](https://cloud.r-project.org/) 
* [RStudio](https://www.rstudio.com/products/rstudio/download/)
* [Integrative Genome Browser (IGV)](https://software.broadinstitute.org/software/igv/download) (optional)

We recommend creating a new "project" in RStudio for this workshop. This will ensure you are working in a fresh working directory. Once you have opened the project, copy and paste the following code to the R console to download necessary files and install packages. During the workshop we will be working through the code in the file "workshop.R".

```{r eval=FALSE}
# Download files for workshop
filenames <- c(
    "workshop.R", 
    "Caenorhabditis_elegans.WBcel235.104.gtf.gz", 
    "Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz",
    "example.bam",
    "example.bam.bai")

download.file(
    paste0("https://monashdatafluency.github.io/r-bioc-2/", filenames),
    filenames)


# Install packages used in workshop
install.packages("BiocManager")
BiocManager::install(c(
    "Biostrings",
    "BSgenome",
    "GenomicRanges",
    "GenomicFeatures",
    "ensembldb",
    "rtracklayer",
    "Rsamtools",
    "Gviz",
    "seqLogo",
    "AnnotationHub"))


# Worm FASTA and GTF files are originally from:
# http://ftp.ensembl.org/pub/release-104/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz
# http://ftp.ensembl.org/pub/release-104/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.104.gtf.gz
```


## Further reading

Most of the packages used in this workshop have vignettes that are worth reading.

For a "tidy" approach to GRanges, check out Stuart Lee's [plyranges](https://bioconductor.org/packages/release/bioc/html/plyranges.html) package. Stuart has also written an introduction to [S4 classes](https://stuartlee.org/2019/07/09/s4-short-guide/).

James MacDonald and Lori Sheppard gave an introduction to Bioconductor annotation resources at Bioc2021: [video](https://www.youtube.com/watch?v=3ABBm7BZszI), [workshop material](https://jmacdon.github.io/Bioc2021Anno/articles/AnnotationWorkshop.html). Good discussion of Bioconductor history.



## Author

This course has been developed for the [Monash Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/bioinformatics) and [Monash Data Fluency](https://www.monash.edu/data-fluency) by Paul Harrison.

<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="CC-BY.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

## Source code

* [GitHub repository](https://github.com/MonashDataFluency/r-bioc-2)

<p style="margin-top: 5em; text-align: right">
<a href="https://www.monash.edu/researchinfrastructure/bioinformatics"><img src="MBP-banner.png" width="675"></a>
</p>
