---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Challenge: sequences

```{r}
rc_seq <- reverseComplement(DNAString("GTAGAGTAATATGGA"))

translate(rc_seq)

translate(subseq(rc_seq,3,11))
# or
translate(rc_seq[3:11])
```


# Challenge: transcript initiation

```{r}
transcripts <- subset(features, type == "transcript" & transcript_biotype == "protein_coding")
flanks <- flank(transcripts, 2, start=TRUE)
flank_seqs <- getSeq(seqs, flanks)
table(flank_seqs)

spans <- resize(flank(transcripts, 2), 4)
# or
spans <- promoters(transcripts, 2, 2)
span_seqs <- getSeq(seqs, spans)
tail(sort(table(span_seqs)), 50)

letter_counts <- consensusMatrix(span_seqs)
probs <- prop.table(letter_counts[1:4,], 2)
seqLogo(probs, ic.scale=FALSE)
seqLogo(probs)
```



# Junk


```{r}

ce_transcripts <- subset(features, type == "transcript" & transcript_biotype == "protein_coding")
ce_ends <- resize(ce_transcripts, width=30, fix="end")
ce_end_seqs <- getSeq(seqs, ce_ends)
names(ce_end_seqs) <- ce_ends$transcript_id

# Which contain the canonical PolyAdenylation Signal (PAS)?
table(vcountPattern("AATAAA", ce_end_seqs))

library(seqLogo)
letter_counts <- consensusMatrix(ce_end_seqs)
probs <- prop.table(letter_counts[1:4,], 2)
seqLogo(probs, ic.scale=FALSE)
seqLogo(probs)



export(ce_end_seqs, "end_seqs.fasta")
# streme --rna --minw 4 --maxw 15 --p end_seqs.fasta
## --rna = look for single-stranded motifs


ce_abouts <- resize(resize(ce_transcripts, width=150, fix="end"), width=300, fix="start")
ce_about_seqs <- getSeq(seqs, ce_abouts)
names(ce_about_seqs) <- ce_abouts$transcript_id
export(ce_about_seqs, "about_seqs.fasta")

letter_counts <- consensusMatrix(ce_about_seqs)
probs <- prop.table(letter_counts[1:4,], 2)
seqLogo(probs, ic.scale=FALSE)
seqLogo(probs)

# streme --rna --minw 4 --maxw 15 --p about_seqs.fasta --o streme_about_out

# centrimo about_seqs.fasta streme_out/streme.xml --local --norc
# centrimo about_seqs.fasta streme_about_out/streme.xml --local --norc
# centrimo about_seqs.fasta ~/meme/motif_databases/WORM/uniprobe_worm.meme --local --norc
# centrimo about_seqs.fasta ~/meme/motif_databases/JASPAR/JASPAR2018_CORE_nematodes_non-redundant.meme --local --norc

#ce_transcripts_2 <- subset(features, type == "transcript" & transcript_biotype == "piRNA")

#ce_starts <- resize(flank(ce_transcripts_2, 10), 20)
ce_starts <- resize(ce_transcripts, 4, fix="start")
ce_start_seqs <- getSeq(seqs, ce_starts)
table(ce_start_seqs)

ce_starts <- resize(flank(ce_transcripts, 2), 4)
ce_start_seqs <- getSeq(seqs, ce_starts)
table(ce_start_seqs) %>% sort()
letter_counts <- consensusMatrix(ce_start_seqs)
probs <- prop.table(letter_counts[1:4,], 2)
seqLogo(probs, ic.scale=FALSE)
seqLogo(probs)




words <- toupper(readLines("/usr/share/dict/words"))

while(TRUE) {
    n <- 100000
    a <- 21
    seq <- DNAStringSet(c("x"=paste(sample(c("A","C","G","T"),n*a,replace=TRUE),collapse="")))
    seqs <- getSeq(seq,GRanges("x",IRanges(start=1+a*(seq_len(n)-1),width=a)))
    t1 <- translate(seqs)
    t2 <- translate(subseq(seqs, 3, 11))
    sel <- as.character(t1) %in% words & as.character(t2) %in% words
    if (any(sel)) {
        print( paste(seqs[sel],t1[sel],t2[sel]) )
    }
}


# TCCATATTACTCTAC SILLY HIT
x <- reverseComplement(DNAString("TCCATATTACTCTAC"))
y <- reverseComplement(DNAString("GTAGAGTAATATGGA"))
translate(y)
translate(subseq(y,3,11))

# GGCGCGCTGGCGAATACG GALANT RAGE (3 14)
# TCGGAAGAGAAGGAGAGG SEEKER GREG

```


